### PR TITLE
Batch Transaction

### PR DESCRIPTION
**Pull request by karthikeyan-cere**
Create a batch transaction for native token transfer

### CODE REVIEW

File: scripts/tokenomics-emulation/config.json [hunk 1]
```diff
@@ -36,7 +36,8 @@
       {
         "name": "native-tokens-transfer",
         "amount": 1,
-        "tokens_range": [1, 2]
+        "tokens_range": [1, 2],
+        "batch_count": 2
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:40]: batch size configuration should be 1 per all emulations. Please, move it on 1 level higher.

File: scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts [hunk 2]
```diff
@@ -11,21 +11,35 @@ class NativeTokensTransferEmulation implements IEmulation {
...
-      const transferAmount = _.random(
-        this.config.tokens_range[0],
-        this.config.tokens_range[1]
-      );
-      const sender = this.account.rootAccount;
-      const transfer = await this.network.transfer(
-        sender,
-        destination.ss58Address,
-        transferAmount.toString()
+    const total = this.config.amount;
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts:14]: We need to reuse this logic in different emulations. Move it to a separate class `Batcher`.

File: scripts/tokenomics-emulation/network.ts [hunk 3]
```diff
@@ -46,23 +46,15 @@ class Network {
...
-    return new Promise((res, rej) => {
-      transfer
-        .signAndSend(
-          sender,
-          { nonce: nonce },
-          Network.sendStatusCb.bind(this, res, rej)
-        )
-        .catch((err) => rej(err));
-    });
+    return transfer;
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/network.ts:57]: Let's leave the ability to execute this transaction without batch.

Just move the transaction creation part to a separate function and reuse it here and in batch preparation.

Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/network.ts:57]: Added a new function to sign and send single transaction

File: scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts (hunk 2 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts]: Added a new class Batcher

File: scripts/tokenomics-emulation/config.json [hunk 4]
```diff
@@ -31,12 +31,14 @@
     }
   ],
   "nodes_amount": 4,
+  "batch_count": 2,
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:34]: It is 2 levels. Put it to the emulation object.

File: scripts/tokenomics-emulation/config.json [hunk 5]
```diff
@@ -31,12 +31,14 @@
...
   "emulations": {
     "sequence": [
       {
         "name": "native-tokens-transfer",
         "amount": 1,
-        "tokens_range": [1, 2]
+        "tokens_range": [1, 2],
+        "batch_count": 2
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/config.json:41]: remove it.

File: scripts/tokenomics-emulation/config.json (hunk 1 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/config.json:41]: Moved to higher level and passed as parameter for the emulaions.

File: scripts/tokenomics-emulation/config.json (hunk 4 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/config.json]: Moved in emulation

File: scripts/tokenomics-emulation/config.json (hunk 5 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/config.json]: Removed

File: scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts [hunk 6]
```diff
@@ -2,31 +2,39 @@ import { IEmulation } from "./emulation.interface";
...
-      const sender = this.account.rootAccount;
-      const transfer = await this.network.transfer(
-        sender,
-        destination.ss58Address,
-        transferAmount.toString()
-      );
-      const balance = await this.network.getBalance(destination.ss58Address);
-      console.log(`Balance of ${destination.ss58Address} is ${balance}`);
-    }
+    const batcher = new Batcher();
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts:16]: Let's create batcher inside main and pass it to the emulations factory.

Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/emulations/transfer-native-token.emulation.ts]: Moved

**Review by AndreiNavoichyk — APPROVED** [review]

File: scripts/tokenomics-emulation/emulations/cere-user-to-app-emulation.ts [hunk 7]
```diff
@@ -1,28 +1,42 @@
...
+    console.log(`Running emulation for user to app cere token transfer`);
+    const sender = this.account.rootAccount;
+    const total = +this.config.amount;
+    await this.batcher.batchProcessing(
+      sender,
+      this.network,
+      total,
+      async () => {
+        const destination = await this.account.generateSrAccount();
+        const tokenValue = await this.config.token_value;
```
Comment (AndreiNavoichyk) [inline scripts/tokenomics-emulation/emulations/cere-user-to-app-emulation.ts:26]: We don't need await here.

**Review by AndreiNavoichyk — APPROVED** [review]

File: scripts/tokenomics-emulation/emulations/cere-user-to-app-emulation.ts (hunk 7 above)
Comment (karthikeyan-cere) [reply scripts/tokenomics-emulation/emulations/cere-user-to-app-emulation.ts]: removed

**Pull request merged**